src/Arcus.WebApi.Correlation/CorrelationMiddleware.cs

```diff
@@ -0,0 +1,133 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Arcus.WebApi.Correlation 
+{
+    /// <summary>
+    /// Correlate the incoming request with the outgoing response by using previously configured options.
+    /// </summary>
+    public class CorrelationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly CorrelationOptions _options;
+        private readonly ILogger<CorrelationMiddleware> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationMiddleware"/> class.
+        /// </summary>
+        /// <param name="next">The next functionality in the request pipeline to be executed.</param>
+        /// <param name="options">The options controlling how the correlation should happen.</param>
+        /// <param name="logger">The logger to trace diagnostic messages during the correlation.</param>
+        /// <exception cref="ArgumentNullException">When any of the parameters are <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">When the <paramref name="options"/> doesn't contain a non-<c>null</c> <see cref="IOptions{TOptions}.Value"/></exception>
+        public CorrelationMiddleware(
+            RequestDelegate next,
+            IOptions<CorrelationOptions> options,
+            ILogger<CorrelationMiddleware> logger)
+        {
+            Guard.NotNull(next, nameof(next), "Requires a continuation delegate");
+            Guard.NotNull(options, nameof(options), "Requires a set of correlation options to manipulate how the correlation should happen");
+            Guard.NotNull(logger, nameof(logger), "Requires a logging implementation to trace diagnostic messages during the correlation");
+            Guard.For<ArgumentException>(() => options.Value is null, "Requires a set of correlation options to manipulate how the correlation should happen");
+
+            _next = next;
+            _options = options.Value;
+            _logger = logger;
+        }
+        
+        /// <summary>Request handling method.</summary>
+        /// <param name="httpContext">The <see cref="T:Microsoft.AspNetCore.Http.HttpContext" /> for the current request.</param>
+        /// <returns>A <see cref="T:System.Threading.Tasks.Task" /> that represents the execution of this middleware.</returns>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            Guard.NotNull(httpContext, nameof(httpContext));
+            Guard.For<ArgumentException>(() => httpContext.Response is null, "Requires a 'Response'");
+            Guard.For<ArgumentException>(() => httpContext.Response.Headers is null, "Requires a 'Response' object with headers");
+
+            if (httpContext.Request.Headers.TryGetValue(_options.Transaction.HeaderName, out StringValues transactionIds))
+            {
+                if (!_options.Transaction.AllowInRequest)
```

Comment:
Thinking out loud:

Do we want to check if it's allowed or just if we need to take it into account?

In other words, do we block the header if specified or do we just ignore it?

Reply:
If we ignore it; I wouldn't call it `Allow...` since that would mean it _would_ block. No?

Reply:
Yes, my question is more from a functional standpoint - What would you like?

But blocking is ok for me.

Reply:
Yeah, I like it when we shortcut it if we 'don't allow' such headers.

Reply:
Sounds good to me